ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/translator/opconventer/HiveGBOpConvUtil.java

```diff
@@ -1028,32 +1028,65 @@ private static OpAttr genReduceSideGB1NoMapGB(OpAttr inputOpAf, GBInfo gbInfo,
-      ArrayList<ExprNodeDesc> aggParameters = new ArrayList<ExprNodeDesc>();
+      final UDAFAttrs udafAttr = gbInfo.udafAttrs.get(i);
+      final List<ColumnInfo> paramColInfo = new ArrayList<>(udafAttr.udafParams.size());
+      final List<String> distinctColNames = new ArrayList<>(udafAttr.udafParams.size());
 
-      ColumnInfo rsUDAFParamColInfo;
-      ExprNodeDesc udafParam;
-      ExprNodeDesc constantPropDistinctUDAFParam;
       for (int j = 0; j < udafAttr.udafParams.size(); j++) {
-        int argPos = getColInfoPos(udafAttr.udafParams.get(j), gbInfo);
-        rsUDAFParamColInfo = rsColInfoLst.get(argPos);
-        String rsUDAFParamName = rsUDAFParamColInfo.getInternalName();
+        final int argPos = getColInfoPos(udafAttr.udafParams.get(j), gbInfo);
+        final ColumnInfo rsUDAFParamColInfo = rsColInfoLst.get(argPos);
+        paramColInfo.add(rsUDAFParamColInfo);
 
+        final String distinctColumnName;
         if (udafAttr.isDistinctUDAF && lastReduceKeyColName != null) {
-          rsUDAFParamName = Utilities.ReduceField.KEY.name() + "." + lastReduceKeyColName
+          distinctColumnName = Utilities.ReduceField.KEY.name() + "." + lastReduceKeyColName
                   + ":" + numDistinctUDFs + "." + SemanticAnalyzer.getColumnInternalName(j);
+          distinctColumnMapping.putIfAbsent(rsUDAFParamColInfo, distinctColumnName);
+        } else {
+          distinctColumnName = null;
+        }
+        distinctColNames.add(distinctColumnName);
+      }
+
+      paramColInfoTable.add(paramColInfo);
+      distinctColumnNameTable.add(distinctColNames);
+
+      if(udafAttr.isDistinctUDAF) {
+        numDistinctUDFs++;
+      }
+    }
+
+    // the positions in rsColInfoLst are as follows
+    // --grpkey--,--distkey--,--values--
+    // but distUDAF may be before/after some non-distUDAF,
+    // i.e., their positions can be mixed.
+    // so for all UDAF we first check to see if it is groupby key, if not is it distinct key
+    // if not it should be value
+    final Map<Integer, List<ExprNodeDesc>> indexToParameter = new TreeMap<>();
+    for (int i = 0; i < paramColInfoTable.size(); i++) {
+      final ArrayList<ExprNodeDesc> aggParameters = new ArrayList<>();
+
+      for (int j = 0; j < paramColInfoTable.get(i).size(); j++) {
+        final ColumnInfo rsUDAFParamColInfo = paramColInfoTable.get(i).get(j);
+
+        final String rsUDAFParamName;
+        if (distinctColumnNameTable.get(i).get(j) != null) {
+          rsUDAFParamName = distinctColumnNameTable.get(i).get(j);
+        } else if (distinctColumnMapping.get(rsUDAFParamColInfo) != null) {
+          // This UDAF is not labeled with DISTINCT, but it refers to a DISTINCT key.
+          // The original internal name is already obsolete as any DISTINCT keys are renamed.
+          rsUDAFParamName = distinctColumnMapping.get(rsUDAFParamColInfo);
+        } else {
+          rsUDAFParamName = rsUDAFParamColInfo.getInternalName();
```

Comment:
As far as I tested, `getInternalName` retains `KEY.col0:{index of RS keys}.col0`. It conflicts with the name generated by the following logic, and a non-distinct UDAF can refer to a wrong column.

```
          distinctColumnName = Utilities.ReduceField.KEY.name() + "." + lastReduceKeyColName
                  + ":" + numDistinctUDFs + "." + SemanticAnalyzer.getColumnInternalName(j);
```

Non-Return Path resolves the correct column name by letting RowResolver resolve the correct name before creating `ExprNodeColumnDesc`. This PR would implement the logic by traversing params with two paths.
Why do we need the two paths? That's because the non-distinct UDAF may refer to the input of DISTINCT UDAF which comes later like below.

```
SELECT
  SUM(DISTINCT col1),
  COUNT(DISTINCT col1),
  SUM(col2), -- This has to refer to the key for `SUM(DISTINCT col2)`
  MAX(DISTINCT col1),
  SUM(DISTINCT col2),
  MIN(DISTINCT col1)
FROM test;
```

Note that this method is used only when `hive.map.aggr` is disabled. With it enabled, non-distinct UDAFs don't refer to DISTINCT keys as they are deduplicated.

Reply:
@okumin 
I'm not familiar with the logic of distinct key generation. IIUC in this case some of these keys are conflicting with non-distinct keys.

Could you please elaborate distinct key generation logic?
Why can we construct keys like this
```
distinctColumnName = Utilities.ReduceField.KEY.name() + "." + lastReduceKeyColName + ":" + numDistinctUDFs + "." + SemanticAnalyzer.getColumnInternalName(j);
```
I was checking the legacy non-cbo return patch code, it also has this
https://github.com/apache/hive/blob/a1420ed816c315d98be7ebf05cdc3ba139a68643/ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java#L5372-L5379
but I still don't understand why can we reference columns which are not coming any schema or what is the schema which contains these generated keys?

Reply:
Thanks for explanation.

Reply:
BTW, HIVE-474 and HIVE-537 explains why RSO + GBO use UNION to express multiple DISTINCTs.

> Why GroupByOperator needs to rename the source column for UNION?

[Reading the primitive change](https://github.com/apache/hive/commit/ed3f0b1983edcdfdc2dbd2d3862d385cbb5c04ae#diff-7d3ff1b241592b075dc3b4afcc01d068e58799c810fc5085815ad6bf8ef5e6a9R609), I guess the operator uses `position of UDAF` of `KEY.{the final output col name}:{position of UDAF}.{position of arguments of the UDAF}` in order to identify if it needs to be DISTINCTed or not while being processed.